### PR TITLE
fix: project name restriction

### PR DESCRIPTION
**Pull request by yuqizhou77**
- Change the restriction of app name. It will allow almost all characters except "<>:\\?/*|\u0000-\u001F and cannot end with dot or space. And it has to start with a letter, and contain at least two alphanumeric characters, and the length should be less than 30. (Same as what we have today).
- AppName in projectSettings will be the original app name that may include special characters
- Remove non-alphanumeric characters in appName in plugins when provisioning resources.
- Update templates for C# projects to use safeProjectName for namespaces and projectName for file name. Related PR: https://devdiv.visualstudio.com/InternalTools/_git/VSProject/pullrequest/404491
- E2E test:
    - https://github.com/OfficeDev/TeamsFx/actions/runs/2506449140 original one
    - https://github.com/OfficeDev/TeamsFx/actions/runs/2507343488 projectName with dot

E2E TEST: na

### CODE REVIEW

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts
Comment (hund030) [inline packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:116]: If we add tab by `addFeature` command, `ctx.answers` may not contain app name. I think we should get app name from project settings.

File: packages/fx-core/src/core/question.ts
Comment (yuqizhou77) [inline packages/fx-core/src/core/question.ts:83]: @MuyangAmigo Hi Junjie, we need to update the value of this string. The current value is "Application name must start with a letter and can only contain letters and digits."

File: packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts
Comment (yuqizhou77) [reply packages/fx-core/src/plugins/resource/frontend/dotnet/plugin.ts:116]: currently vs does not have "addFeature" command. need to sync with the team to understand the current process of that feature.

File: packages/fx-core/src/plugins/resource/aad/utils/configs.ts
Comment (SLdragon) [inline packages/fx-core/src/plugins/resource/aad/utils/configs.ts:271]: Seems multiple places has these code, consider extract to a function:
```
    const displayName: string = convertToAlphanumericOnly(ctx.projectSetting.appName);
    if (displayName) {
      this.displayName = displayName.substr(0, Constants.aadAppMaxLength) as string;
    } else {
      throw ResultFactory.SystemError(
        GetConfigError.name,
        GetConfigError.message(Errors.GetDisplayNameError[0])
      );
    }
```

**Review by hund030 — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/aad/utils/configs.ts
Comment (yuqizhou77) [reply packages/fx-core/src/plugins/resource/aad/utils/configs.ts]: Fixed

File: packages/cli/tests/e2e/commonUtils.ts
Comment (nliu-ms) [inline packages/cli/tests/e2e/commonUtils.ts:98]: Just wondering will this happen: the appname contains only special characters, and after replacing it becomes an empty string.

Comment (nliu-ms) [reply packages/cli/tests/e2e/commonUtils.ts:98]: Never mind, it should start from a letter.

**Review by nliu-ms — DISMISSED** [review]

File: packages/fx-core/src/core/question.ts
Comment (MuyangAmigo) [reply packages/fx-core/src/core/question.ts:83]: Proposed message: `Application name must start with letters and contain at least two letters or digits. It can not contain specific ASCII characters.`

**Review by hund030 — APPROVED** [review]

**Review by KennethBWSong — DISMISSED** [review]

**Review by yukun-dong — DISMISSED** [review]

**Review by nliu-ms — APPROVED** [review]

**Review by SLdragon — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/spfx/plugin.ts
Comment (HuihuiWu-Microsoft) [inline packages/fx-core/src/plugins/resource/spfx/plugin.ts:110]: If you don't want to dismiss all the existing approvals, I'll remove the comment later in next PR.

**Review by HuihuiWu-Microsoft — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/spfx/plugin.ts
Comment (yuqizhou77) [reply packages/fx-core/src/plugins/resource/spfx/plugin.ts:110]: Thanks for catching this! I will have another PR to remove this comment after checking this one in :)

Comment (yuqizhou77) [reply packages/fx-core/src/plugins/resource/spfx/plugin.ts]: Fixed

**Review by jayzhang — APPROVED** [review]

**Review by HuihuiWu-Microsoft — APPROVED** [review]

**Review by LongOddCode — APPROVED** [review]

**Review by tecton — APPROVED** [review]

**Review by chagong — APPROVED** [review]

**Review by nliu-ms — APPROVED** [review]

**Review by yukun-dong — APPROVED** [review]

**Review by Siglud — APPROVED** [review]

**Review by 1yefuwang1 — APPROVED** [review]

**Review by XiaofuHuang — APPROVED** [review]

**Review by dooriya — APPROVED** [review]

**Review by hund030 — APPROVED** [review]

**Pull request merged**
